### DISCUSSION TITLE
`os.spawn`: confused about `env` arg optional-ness

### DISCUSSION THREAD

**Discussion by metaleap** [orbitalquark/textadept#156] [Q&A] [answered]
The `env` arg of `os.spawn` is described as optional. But it seems it cannot be `nil` or a process-exit event will actually end up in `stderr_cb` not `exit_cb`.

I can leave it out totally, no problem. Or pass a non-nil env, no problem. But somehow with passing `nil`, all the callbacks are "shifted" or something. I noticed that @orbitalquark in [textadept-lsp](https://github.com/orbitalquark/textadept-lsp) just omits the `env` arg altogether.

But if my script's os.spawn is parameterized by my script's callers/users, that means my script has to do the nil check, then either do the 5-args `os.spawn` or the 6-args one...

Am I missing some Lua language subtlety here? :grin:

**Comment by orbitalquark** [comment, marked as answer]
This is deliberate. You shouldn't pass `nil` for env, but omit it completely. I've found the vast majority of the time, you want an inherited env, so passing `nil` before callbacks gets annoying. It's not a Lua language quirk. You can create a table of args, insert env if needed, and then call `table.unpack()` when calling `os.spawn`. See https://github.com/orbitalquark/textadept/blob/d7b230840e6fbc493da8075218d55186d4dab4b6/modules/textadept/run.lua#L178

**Reply by metaleap** [reply]
Alright good to know, was perplexed by the args being apparented "shifted left" though (coming from very non-Lua languages =)
